src/game-coordinator/services/player-behavior-handler.service.ts

```diff
@@ -0,0 +1,54 @@
+import { ConfigurationService } from '@/configuration/services/configuration.service';
+import { GameState } from '@/games/models/game-state';
+import { PlayerConnectionStatus } from '@/games/models/player-connection-status';
+import { GamesService } from '@/games/services/games.service';
+import { PlayerSubstitutionService } from '@/games/services/player-substitution.service';
+import { PlayersService } from '@/players/services/players.service';
+import { Injectable, Logger } from '@nestjs/common';
+import { Cron, CronExpression } from '@nestjs/schedule';
+
+@Injectable()
+export class PlayerBehaviorHandlerService {
+  private logger = new Logger(PlayerBehaviorHandlerService.name);
+
+  constructor(
+    private readonly gamesService: GamesService,
+    private readonly playerSubstitutionService: PlayerSubstitutionService,
+    private readonly playersService: PlayersService,
+    private readonly configurationService: ConfigurationService,
+  ) {}
+
+  @Cron(CronExpression.EVERY_30_SECONDS)
+  async verifyPlayersJoinedGameServer() {
+    const bot = await this.playersService.findBot();
+    const timeout = (await this.configurationService.getTimeToJoinGameServer())
+      .value;
+    const gamesLive = await this.gamesService.getRunningGames();
+    await Promise.all(
+      gamesLive
+        .filter((game) => game.state === GameState.launching)
+        .filter((game) => Boolean(game.lastConfiguredAt))
+        .filter((game) => game.launchedAt.getTime() + timeout < Date.now())
+        .map(async (game) => {
+          return await Promise.all(
+            game
+              .activeSlots()
+              .filter(
+                (slot) =>
+                  slot.connectionStatus === PlayerConnectionStatus.offline,
+              )
+              .map(async (slot) => {
+                this.logger.warn(
```

Comment:
Nit on the warn- is this a warn or a general INFO? I believe info as it does not partain to the server's overall health (and I think logging overall should be INFO if it relates to happy paths- to which this is a happy path as it can happen and it shouldn't set off alarms)

Reply:
I'm not going to hold the PR against this- but I think generally our logging technique should be standardised.

In short, anything that could mean that we've hit undesired code paths or undefined behavior should be FATAL. 

Anything that is third-party related being down and not covered, ERROR.

Application state mismanagement that can be rectified by code (e.g. "self healing") should be WARN.

Any gameplay mechanics or information about non-server related things should be LOG.

I love your debug logs- keep them!! Maybe use TRACE when adding more debug info.

Reply:
Fair point